### PR TITLE
feat: introduce experiment created and updated events

### PR DESCRIPTION
**Pull request by yoganandaness**
## Description

This pull request introduces significant improvements to the backend's event handling and message brokering, especially around experiments and proposals. The main changes include implementing a new function to generate comprehensive experiment message data (including related samples, proposal, and instrument information), updating the RabbitMQ message broker to handle new experiment-related events, and enhancing mock data sources and tests to support these features.

## Motivation and Context

Experiment event support and messaging

## How Has This Been Tested

Attached Unit tests

## Changes

**Enhancements to message brokering and event handling:**

* Added a new `getExperimentMessageData` function in `messageBroker.ts` to construct detailed JSON messages for experiments, including associated samples, proposal, and instrument data.
* Updated the RabbitMQ event handler to publish messages for `EXPERIMENT_CREATED`, `EXPERIMENT_UPDATED`, and `EXPERIMENT_ESF_SUBMITTED` events using the new experiment message data structure.
* Improved proposal message data to include associated samples, and ensured proposals must exist before sending messages. 

**Testing improvements:**

* Added comprehensive tests for `getExperimentMessageData` to verify inclusion of instrument, samples, and proposer data, and to handle missing instruments.
* Introduced a new test suite for the experiment safety workflow event handler, covering edge cases and status change event publishing.

## Tests included/Docs Updated?

- [x] I have added tests to cover my changes.
- [ ] All relevant doc has been updated

### CODE REVIEW

File: apps/backend/src/eventHandlers/messageBroker.ts [hunk 1]
```diff
@@ -68,6 +69,18 @@ type ProposalMessageData = {
   shortCode: string;
   title: string;
   submitted: boolean;
+  samples?: { id: number; title: string }[];
```
Comment (jekabs-karklins) [inline apps/backend/src/eventHandlers/messageBroker.ts]: suggestion: Can we do a simple Pick<Sample, 'id' | 'title'> here. This would help if anyone refactors in the future

File: apps/backend/src/eventHandlers/messageBroker.ts [hunk 2]
```diff
@@ -68,6 +69,18 @@ type ProposalMessageData = {
...
+};
+
+type ExperimentMessageData = {
+  experimentId: string;
+  experimentPk: number;
+  startsAt: Date;
+  endsAt: Date;
+  status: string;
+  proposal?: ProposalMessageData;
+  samples?: { id: number; title: string }[];
```
Comment (jekabs-karklins) [inline apps/backend/src/eventHandlers/messageBroker.ts]: suggestion: ditto

File: apps/backend/src/eventHandlers/messageBroker.ts [hunk 3]
```diff
@@ -187,6 +204,14 @@ export const getProposalMessageData = async (proposal: Proposal) => {
     visitors: visitorsWithInstitution.map(mapUserWithInstitutionToMember),
     newStatus: proposalStatus?.shortCode,
     submitted: proposal.submitted,
+    samples: (
+      await sampleDataSource.getSamples({
+        filter: { proposalPk: proposal.primaryKey },
+      })
+    ).map((sample) => ({
```
Comment (jekabs-karklins) [inline apps/backend/src/eventHandlers/messageBroker.ts:212]: suggestion: ditto

**Review by jekabs-karklins — APPROVED** [review]
Looks good

**Review by TCMeldrum — APPROVED** [review]

**Comment by jekabs-karklins** [comment]
I think this could be merged
